Topic type completion & Service completion

sven-hoek opened the pull request:
Regarding https://github.com/kpbaks/ros2.fish/issues/9
Basically see commit titles. Let me know if you disagree with adding comments etc. or if this should be split into _add service name completion_ and _add topic type completion_.

Service name completion (line 665 to 666) is straightforward and analog to topic name completion, which existed beforehand.

Completing the topic type was a bit trickier with regards to getting the condition right. I'm still not fully happy as command line options that take an argument (e.g. `--rate 10`) make up 2 tokens, one of which isn't starting with a `-` (I remove any tokens starting with a dash) and would invalidate the condition. Thinking about it, it could be better to simply check, whether the last token of the command line is a topic from`(ros2 topic list)`.  Maybe I will do that but not today. Let me know if you have any suggestions, though.

It works as it is but could be a bit more robust. Once it is, something similar could be done for services. And using `ros2 interface proto <interface type>`, adding completion for filling out the message/service request could probably be somewhat easy.

sven-hoek committed:
Rename to __fish_ros2_all_subcommands

It's more in line with __fish_ros2_all_commands

sven-hoek committed:
Add some function description commands

sven-hoek committed:
Remove an unused and an commented out function

sven-hoek committed:
Add service completions for call and info

sven-hoek committed:
Add topic type completion for pub

kpbaks commented on completions/ros2.fish:
```diff
@@ -3,9 +3,25 @@ set -l C complete --command ros2
+# Remove tokens starting with `-`
+function __fish_remove_options_from_commandline
+    for i in (seq (count $argv))
+        set -l first_letter "$(string sub --length 1 -- $argv[$i])"
```
in fish you can do `(command)` instead of `"$(command)"`

kpbaks commented on completions/ros2.fish:
```diff
@@ -3,9 +3,25 @@ set -l C complete --command ros2
 
+# Remove tokens starting with `-`
+function __fish_remove_options_from_commandline
+    for i in (seq (count $argv))
```
All variables in fish are lists an can be iterated over with `for arg in $argv`

kpbaks commented:
Looks good.
I will need to get a ros2 environment up and running in a virtual environment to test it.

sven-hoek committed:
Simplify message type completion

- Removing "option" tokens from the command line was not considering option-values (e.g. --option value), so the removal was faulty
- Instead, it makes more sense to look at the last token and check if it's a topic
- Also, it makes it easier to check, whether a topic name should be suggested by simply checking if nothing on the command line is a topic already

sven-hoek committed:
Remove duplicate function

sven-hoek committed:
Add completion for service type

sven-hoek committed:
Use $fish_ros variable

sven-hoek committed:
Add completion for lifcycle nodes and their transitions

sven-hoek committed:
Add completions for message and service request  content

Once a topic name and message type (or service name and type) have been entered in the command line, pressing TAB will insert a string with all fields of the message and their default values, making it easy to publish even complex messages.
Note that below fish version 4.x, this will insert an ugly string with lots of escaping backlashes. As it doesn't really break anything, I didn't change the requirement to be higher than 4, only removed an error when using a larger major version with the minor being smaller than 6.

sven-hoek commented:
I added more features. I wanted to create separate PRs for them but they kinda build upon what's in this one. Let me know if they should be separated anyway.
- Completion of service type when doing `ros2 service call <service name>`
- Completion of lifecycle nodes when doing `ros2 lifecycle get|set|list`
- Completion of lifecycle transitions when doing `ros2 lifecycle set <node name>`
![image](https://github.com/user-attachments/assets/bb4aa99a-27e6-4978-90ce-c355689d3c85)
- Completion of the message/service request content, making it easy to publish/call with complex message/request types. This requires fish version 4.x to work nicely, otherwise the string is full of escaping-backlashes. I fixed the version requirement-check to also accept majors larger than 3 but as it still kinda works, I didn't make it a requirement.
![image](https://github.com/user-attachments/assets/8822fa6e-df33-4a96-987d-31fea21ddb20)

sven-hoek committed:
Add completion for launch arguments

kpbaks commented:
I like it!
I won't be available tomorrow and over the weekend, so I will first have a chance to look at it properly next week.

sven-hoek committed:
Merge branch 'main' into topic-type-completion
